lib/dsl/chev_grammar.js

```diff
@@ -114,20 +131,135 @@ createToken({ name: 'INTEGER', pattern: /-?\d+/ });
+    // like in pegjs. This capability is useful for partial parsing, e.g.:
+    // 1. Code snippets
+    // 2. Incremental parsing of only the changed parts of an active Editor.
+    // 3. writing Unit tests for micro code samples.
+    $.RULE('prog', () => {
+      $.OR([
+        { ALT: () => { $.SUBRULE($.constantDecl); } },
+        { ALT: () => { $.SUBRULE($.entityDecl); } }
+      ]);
+    });
+
+    $.RULE('constantDecl', () => {
+      $.CONSUME(t.NAME);
+      $.CONSUME(t.EQUALS);
+      $.CONSUME(t.INTEGER);
+    });
+
+    $.RULE('entityDecl', () => {
+      $.CONSUME(t.ENTITY);
+      $.CONSUME(t.NAME);
+
+      $.OPTION(() => {
+        $.SUBRULE($.entityTableNameDecl);
+      });
+
+      // the "2" suffix is a quirk of Chevrotain, more details:
+      // https://github.com/SAP/chevrotain/blob/master/docs/faq.md#-why-are-the-unique-numerical-suffixes-consume1consume2-needed-for-the-dsl-rules
+      $.OPTION2(() => {
+        $.SUBRULE($.entityBody);
+      });
+    });
+
+    $.RULE('entityTableNameDecl', () => {
+      $.CONSUME(t.LPAREN);
+      $.CONSUME(t.NAME);
+      $.CONSUME(t.RPAREN);
+    });
+
+    $.RULE('entityBody', () => {
+      $.CONSUME(t.LCURLY);
+      $.AT_LEAST_ONE(() => {
+        $.SUBRULE($.fieldDec);
+      });
+      $.CONSUME(t.RCURLY);
+    });
+
+    $.RULE('fieldDec', () => {
+      $.CONSUME(t.NAME);
+      $.SUBRULE($.type);
+      // Short form for: "(X(,X)*)?"
+      $.MANY_SEP({
+        SEP: t.COMMA,
+        DEF: () => {
+          $.SUBRULE($.validation);
+        }
+      });
+      $.CONSUME(t.RCURLY);
+    });
+
+    $.RULE('type', () => {
+      $.CONSUME($.NAME);
+    });
+
+    $.RULE('validation', () => {
+      $.OR([
+        { ALT: () => { $.CONSUME(t.REQUIRED); } },
+        { ALT: () => { $.SUBRULE($.minMaxValidation); } },
+        { ALT: () => { $.SUBRULE($.pattern); } }
+      ]);
+    });
+
+    $.RULE('minMaxValidation', () => {
+      // HIGHLIGHT:
+      // Note that "MIN_MAX_KEYWORD" is an abstract token and could match 6 different concrete token types
+      $.CONSUME(t.MIN_MAX_KEYWORD);
+      $.CONSUME(t.LPAREN);
+      $.OR([
+        { ALT: () => { $.CONSUME(t.INTEGER); } },
+        { ALT: () => { $.CONSUME(t.NAME); } }
+      ]);
+      $.CONSUME(t.RPAREN);
+    });
+
+    $.RULE('pattern', () => {
+      $.CONSUME(t.PATTERN);
+      $.CONSUME(t.LPAREN);
+      // HIGHLIGHT:
+      // With Chevrotain the grammar can be debugged directly by using good old fashioned breakpoints.
+      // No need to to try and figure out a 10,000 lines generated file, or worse not even have that
+      // if we would be use some JS combinator.
+      // debugger; 
+      $.CONSUME(t.REGEX);
+      $.CONSUME(t.RPAREN);
+    });
+
+    // very important to call this after all the rules have been defined.
+    // otherwise the parser may not work correctly as it will lack information
+    // derived during the self analysis phase.
+    Parser.performSelfAnalysis(this);
```

Comment:
This is during the object's construction-time. All this. Why not having another way?
One file is enough, but putting everything in the constructor isn't really something I look forward to maintaining, even if the improvement of using using Chevrotain over PegJS is obvious. Why not, for instance, use a factory of some sort (a function that calls other functions to build the parser instance)?

Reply:
> Why not having another way?

### Answer
The syntax I prefer relies on using **class fields** ESNext syntax.
https://github.com/tc39/proposal-class-fields
But this is not yet supported afaik (currently stage 3 proposal).
I suppose Babel will support this at some point:
https://github.com/babel/proposals/issues/12

TypeScript has something similar which already works **now**.
See [this example](https://github.com/SAP/chevrotain/blob/master/examples/implementation_languages/typescript/typescript_json.ts#L69):
This is similar the "official" API I'm aiming for, but may need to wait for ES2018 for that. :(


### Alternative
Anyhow as it is all just plain JavaScript you can define it (mostly) however you want...
An extreme example would be this completely different DSL for specifying Chevrotain grammars
https://github.com/kristianmandrup/chevrotain-rule-dsl

I'm am a bit too tired to think a concrete alternative syntax right now.
But I believe one should be possible even with ES6, perhaps you have a suggestion?
The constraints are:
1. Parser.**performSelfAnalysis** must be called after the rules have been defined.
    * As it relies on side effects of creating the rules.

2. The **RULE** calls must be called in the context of  the parser instance (this).

And if it helps normally you only use a single parser instance and **reset** it's internal state before each use.

### Future / Long term.
There is also an [open issue](https://github.com/SAP/chevrotain/issues/480) for better support of custom APIs for building Chevrotain parsers.
And I'm hoping in the long term to support **three** different API styles (same as Mocha/Chai have different APIs using the same underlying engine). 
1. Low Level Hand-Built style.
2. Combinator Style, fluent DSL.
3. EBNF generator style (Like pegjs).

Reply:
Here is a really quick and dirty factory style hack.
https://github.com/SAP/chevrotain/blob/5235a12da1818aaf2ac075cd4326d46e46da15fc/examples/grammars/json/json.js#L95-L126

And here are the rules defined **outside** the constructor.
https://github.com/SAP/chevrotain/blob/5235a12da1818aaf2ac075cd4326d46e46da15fc/examples/grammars/json/json.js#L129-L180

I don't think this should be part of Chevrotain's official API
as I would rather wait for class fields proposal, but it can be cleaned up and reused 
by end users if needed...

Hope this example demonstrates how due to Chevrotain being a **library**
instead of a code generator makes it much more malleable to customization. 😄